### PR TITLE
add br decompression support

### PR DESCRIPTION
**Pull request by yinwoods**

### CODE REVIEW

**Comment by garydgregory** [comment]
I do not think we want a direct dependency or hard wired support for one of many compression formats.
You might be better off with your app depending on Apache Commons Compress.

**Comment by yinwoods** [comment]
OK, make sense to me. Apache Commons Compress also depends on this dec dependency, I will close this PR.

**Comment by yinwoods** [comment]
Make brotli dependency optional like conscrypt. But not sure this implementation is elegant enough

**Comment by garydgregory** [comment]
IMI, instead, we could have a new Maven module that uses Apache Commons Compress, see my previous comment https://github.com/apache/httpcomponents-client/pull/363#issuecomment-1124487913

**Comment by ok2c** [comment]
@yinwoods Looks pretty good to me. I suppose `ContentCompressionExec` could be optimized to reduce code duplication but it is not a deal breaker for me.

**Review by ok2c — APPROVED** [review]

**Comment by ok2c** [comment]
@yinwoods Something is wrong now. Now the tests fail due to missing br optional dependency, but they should not be. The Brotli support should be wired only when br is available on the classpath. Please take a look.

**Comment by yinwoods** [comment]
> @yinwoods Something is wrong now. Now the tests fail due to missing br optional dependency, but they should not be. The Brotli support should be wired only when br is available on the classpath. Please take a look.

Yes，I'm working on it.

**Comment by yinwoods** [comment]
It's wired, once BrotliInputStreamFactory instantiationed, will throw java.lang.NoClassDefFoundError: org/brotli/dec/BrotliInputStream.

File: pom.xml [hunk 1]
```diff
@@ -65,6 +65,7 @@
     <httpcore.version>5.2-beta1</httpcore.version>
     <log4j.version>2.17.0</log4j.version>
     <commons-codec.version>1.15</commons-codec.version>
+    <dec.version>0.1.2</dec.version>
```
Comment (michael-o) [inline pom.xml:68]: Please rename to brotli. `dec` is a quite generic name.

File: pom.xml [hunk 2]
```diff
@@ -148,6 +149,11 @@
         <artifactId>commons-codec</artifactId>
         <version>${commons-codec.version}</version>
       </dependency>
+      <dependency>
+        <groupId>org.brotli</groupId>
+        <artifactId>dec</artifactId>
```
Comment (michael-o) [inline pom.xml:154]: Very bad choice for an artifactId from the brotli team. (not related to this PR)

**Review by michael-o — CHANGES_REQUESTED** [review]

File: httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/ContentCompressionExec.java [hunk 3]
```diff
@@ -107,6 +125,7 @@ public ContentCompressionExec(final boolean ignoreUnknown) {
      * <ul>
      * <li>gzip - see {@link java.util.zip.GZIPInputStream}</li>
      * <li>deflate - see {@link org.apache.hc.client5.http.entity.DeflateInputStream}</li>
+     * <li>brotli - see {@link org.brotli.dec.BrotliInputStream}</li>
```
Comment (michael-o) [inline httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/ContentCompressionExec.java:128]: The key says `br`, but this `brotli`. Intentional?

File: httpclient5/src/test/java/org/apache/hc/client5/http/entity/TestBrotli.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * Brotli decompression test implemented by request with specified response encoding br
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testDecompressionWithBrotli() throws Exception {
+        final HttpClient client = HttpClientBuilder.create().build();
+        final HttpGet get = new HttpGet("https://www.baidu.com");
```
Comment (michael-o) [inline httpclient5/src/test/java/org/apache/hc/client5/http/entity/TestBrotli.java:48]: I don't accept this test. It requires outbound access which is brittle. TEst should be selfcontained.

**Review by michael-o — CHANGES_REQUESTED** [review]

File: httpclient5/src/test/java/org/apache/hc/client5/http/entity/TestBrotli.java (hunk 4 above)
Comment (ok2c) [reply httpclient5/src/test/java/org/apache/hc/client5/http/entity/TestBrotli.java:48]: @michael-o Good catch. I overlooked it. @yinwoods Please either use our embedded test server or unit test `ContentCompressionExec` which would be preferred.

Comment (yinwoods) [reply httpclient5/src/test/java/org/apache/hc/client5/http/entity/TestBrotli.java:48]: OK, I'll reconsider the unit test. Thanks for the reviewing.

File: httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/ContentCompressionExec.java (hunk 3 above)
Comment (yinwoods) [reply httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/ContentCompressionExec.java:128]: Yes, br is abbreviations of brotli. When in HTTP header Accept-Encoding, br is used.

**Comment by michael-o** [comment]
I can't even pass the current tests, they hang with:
```

```
Because they create too many threads (thread dump of the hanging Surefire Booter) JVM:
```
$ grep "I/O server dispatch-" out  | cut -d ' ' -f -3
"I/O server dispatch-1220"
"I/O server dispatch-1219"
"I/O server dispatch-1218"
"I/O server dispatch-1217"
"I/O server dispatch-1216"
"I/O server dispatch-1215"
"I/O server dispatch-1214"
"I/O server dispatch-1213"
"I/O server dispatch-1212"
"I/O server dispatch-1211"
"I/O server dispatch-1210"
"I/O server dispatch-1209"
"I/O server dispatch-1208"
"I/O server dispatch-1207"
"I/O server dispatch-1206"
"I/O server dispatch-1205"
"I/O server dispatch-1204"
"I/O server dispatch-1203"
"I/O server dispatch-1202"
"I/O server dispatch-1201"
"I/O server dispatch-860"
"I/O server dispatch-859"
"I/O server dispatch-858"
"I/O server dispatch-857"
"I/O server dispatch-856"
"I/O server dispatch-855"
"I/O server dispatch-854"
"I/O server dispatch-853"
"I/O server dispatch-852"
"I/O server dispatch-851"
"I/O server dispatch-850"
"I/O server dispatch-849"
"I/O server dispatch-848"
"I/O server dispatch-847"
"I/O server dispatch-846"
"I/O server dispatch-845"
"I/O server dispatch-844"
"I/O server dispatch-843"
"I/O server dispatch-842"
"I/O server dispatch-841"

```
My OS limit is:
```
$ sysctl kern.threads.max_threads_per_proc
kern.threads.max_threads_per_proc: 1500
```

and I guess this is more than enough for testing, no?

**Comment by ok2c** [comment]
@michael-o Tests pass on Travis and Jenkins. Are you running FreeBSD? If you are getting the same OOM with the current master, can we deal with the issue after merging this PR?

**Comment by michael-o** [comment]
> @michael-o Tests pass on Travis and Jenkins. Are you running FreeBSD? If you are getting the same OOM with the current master, can we deal with the issue after merging this PR?

Yes, running on FreeBSD and OOM is on master, I didn't even get so far to try this PR.

**Review by michael-o — COMMENTED** [review]
Looks reasonable to me.

**Pull request merged**